vcs/git.go

```diff
@@ -47,37 +62,79 @@ func (g *GitDriver) HeadRev(dir string) (string, error) {
 	return strings.TrimSpace(buf.String()), cmd.Wait()
 }
 
-func run(desc, dir, cmd string, args ...string) error {
+func detectRef(dir string) string {
+	output, err := run("git show remote info", dir,
+		"git",
+		"remote",
+		"show",
+		"origin",
+	)
+
+	if err != nil {
+		log.Printf(
+			"error occured when fetching info to determine target ref in %s: %s. Will fall back to default ref %s",
+			dir,
+			err,
+			defaultRef,
+		)
+		return ""
+	}
+
+	matches := headBranchRegexp.FindStringSubmatch(output)
+	if len(matches) != 2 {
+		log.Printf(
+			"could not determine target ref in %s. Will fall back to default ref %s",
+			dir,
+			defaultRef,
+		)
+		return ""
+	}
+
+	return matches[1]
+}
+
+func run(desc, dir, cmd string, args ...string) (string, error) {
 	c := exec.Command(cmd, args...)
 	c.Dir = dir
-	if out, err := c.CombinedOutput(); err != nil {
+	out, err := c.CombinedOutput()
+	if err != nil {
 		log.Printf(
 			"Failed to %s %s, see output below\n%sContinuing...",
 			desc,
 			dir,
 			out)
-		return err
 	}
-	return nil
+
+	return string(out), nil
 }
 
 func (g *GitDriver) Pull(dir string) (string, error) {
-	if err := run("git fetch", dir,
+	var targetRef string
+	if len(g.Ref) > 0 {
```

Comment:
For consistency, empty string comparisons should be `g.Ref == ""`

Reply:
Will do - thanks!